### PR TITLE
Add Configurability

### PR DESCRIPTION
**Pull request by mnguo98**
1. Configurability changes
2. Read through `dependencies.yaml` of repos with a `deploy` directory
3. Handles subdirectories within the repos folder

### CODE REVIEW

**Comment by mnguo98** [comment]
Thought came to me right after I made this PR, but I'm not sure if poolID and clientID should be configurable by the client.

File: src/getClient.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const readlineSync = require('readline-sync');
+const JupiterOneClient = require('@jupiterone/jupiterone-client-nodejs');
```
Comment (aiwilliams) [inline src/getClient.ts:1]: Is there a reason these cannot be `import` statements?

File: src/getClient.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+let J1_ACCOUNT, J1_USERNAME, J1_PASSWORD, J1_API_TOKEN;
```
Comment (aiwilliams) [inline src/getClient.ts]: I would expect these to be immutable constants because of the all caps names. I think you could just use local vars in `getClient`, such as `const accountId = clientInput[0] === '' ? process.env.J1_ACCOUNT : clientInput[0]`, or even avoid modifying the stack:

```
return new JupiterOneClient({
     account: clientInput[0] === '' ? process.env.J1_ACCOUNT : clientInput[0],
     ...
   }).init();
```

Also, I wonder why the need to check for empty strings, but I've not dug in more to find out.

File: src/getdependencies.ts [hunk 3]
```diff
@@ -1,24 +1,44 @@
 import { error } from 'console';
+const readYaml =  require('read-yaml');
```
Comment (aiwilliams) [inline src/getdependencies.ts]: Again, why not `import ...`?

File: src/getdependencies.ts [hunk 4]
```diff
@@ -1,24 +1,44 @@
...
-let readJson;
+let readJson, parseYaml;
 let dependencies = [];
 
-export function getDependencies(repo: string, packageScope: string[]) {
+export function getDependencies(repo: string, packageScope: string[], repoMap: Map<string,string>) {
   try {
-    let path = '../repos/' + repo + '/package.json';
-
+    let path = '../' + repoMap.get(repo) + '/' + repo + '/package.json';
```
Comment (aiwilliams) [inline src/getdependencies.ts]: I don't see `path` modified. `const` is best when it isn't modified. `let` should be rare. With `let`, I have to understand what could cause that state to change.

File: src/getdependencies.ts [hunk 5]
```diff
@@ -1,24 +1,44 @@
...
     dependencies = [];
```
Comment (aiwilliams) [inline src/getdependencies.ts:9]: I think it would be good to declare this as `const` as well. In fact, it seems that adding `tslint` with recommended settings would be worthwhile for this project.

File: src/index.ts [hunk 6]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
-    clientId: process.env.J1_CLIENT_ID,
-    accessToken: process.env.J1_API_TOKEN,
-    dev: process.env.J1_DEV || false
-  }).init();
+  const clientInput = [];
+  const pathToRepos = readlineSync.question('Input path to directory with repos (relative to root directory): ');
+  clientInput.push(readlineSync.question('Input account. If using a .env file, continue without input: '));
+  clientInput.push(readlineSync.question('Input username. If using a .env file, continue without input: '));
+  clientInput.push(readlineSync.question('Input password. If using a .env file, continue without input: '));
+  clientInput.push(readlineSync.question('Input access token. If using a .env file, continue without input: '));
```
Comment (aiwilliams) [inline src/index.ts]: Ah, now I see how values could be empty strings. I'm a fan of dealing with that here in the readline codes, and passing `undefined` as values to `getClient`. Also, would be great to go for `const clientInput = { account: sanitizedValue, ... }` instead of a positional array.

File: src/index.ts [hunk 7]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
+            failure++;
+            if (missingDeps.indexOf(dep) === -1) {
+              missingDeps.push(dep + ' (' + repoName + ').');
+            }
+          }
+        }
+      } else {
+        console.log('Repo has no dependencies of requested scopes');
+      }
+    if (deployDepsList != undefined) {
```
Comment (aiwilliams) [inline src/index.ts]: Would be nice to break this file up into well-name functions, this is getting hard to follow.

**Review by aiwilliams — COMMENTED** [review]
I've added some thoughts, but did not read every line. Consider many of them as general guidelines, not necessarily tasks to win my approval 😄 I realize this is a script, not production code per-se.

File: src/getdependencies.ts (hunk 4 above)
Comment (austin-rausch) [inline src/getdependencies.ts]: You should probably use the path module for this: https://nodejs.org/docs/latest-v10.x/api/path.html#path_path_resolve_paths

This should help it play nicely on windows PCs (which I'm sure many of our customers use)

File: src/getdependencies.ts [hunk 8]
```diff
@@ -1,24 +1,44 @@
...
+          if (j.includes(scope)) {
+            dependencies.push(j + ': ' + readJson[j]);
+          }
         }
       }
     }
     return dependencies;
   } catch (error) {
-    console.log('Repo does not have a package.json');
+    console.log('*** Repo does not have a package.json ***');
```
Comment (austin-rausch) [inline src/getdependencies.ts:32]: Would be helpful to include the repo path or name here for debugging purposes.

File: src/getdependencies.ts [hunk 9]
```diff
@@ -1,24 +1,44 @@
...
+  }
+}
+
+export function getDependenciesYaml(repo: string, repoMap: Map<string,string>) {
+  try {
+    let path = repoMap.get(repo) + '/' + repo + '/deploy/dependencies.yaml';
```
Comment (austin-rausch) [inline src/getdependencies.ts]: Another place the `path` module would be helpful.

File: src/getrepos.ts [hunk 10]
```diff
@@ -1,31 +1,37 @@
...
+  const repos = fs.readdirSync(repoPath);
+  const j1Client = await getClient(clientInput);
 
   for (const repo of repos) {
     const repoID = await j1Client.queryV1(
       `Find CodeRepo with name='${repo}'`,
     );
     if (repoID.length === 0) {
-      console.log('Could not query Repo (' + repo + ').');
+      let newRepoPath = repoPath.substring(repoPath.length) === '/' ? repoPath + repo : repoPath + '/' + repo;
```
Comment (austin-rausch) [inline src/getrepos.ts]: Another place where the `path` module would be helpful.

File: src/getrepos.ts [hunk 11]
```diff
@@ -1,31 +1,37 @@
...
+      let newRepos = fs.readdirSync(newRepoPath);
+      let subdir = false;
+      for (const nrepo of newRepos) {
+        const lol = await j1Client.queryV1(`Find CodeRepo with name='${nrepo}'`);
```
Comment (austin-rausch) [inline src/getrepos.ts]: Why you `lol` at query. Pls don't bully.

File: src/index.ts (hunk 7 above)
Comment (mnguo98) [reply src/index.ts]: 😅 Sorry about that. I'll work on that

Comment (austin-rausch) [inline src/index.ts]: Some will disagree with this comment, but I don't think its ever a good practice to do `!=` instead of `!==` in JavaScript, if you need to check for not equal to `undefined` or `null` do it explicitly, E.G.
```
if (deployDepsList !== undefined || deployDepsList !== null) {
...
```

File: src/getClient.ts (hunk 1 above)
Comment (mnguo98) [reply src/getClient.ts:1]: It was force of habit, my bad. I'll fix all of 'em.

File: src/index.ts [hunk 12]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
-            mainRepo[0].entity._id,
-            depRepo[0].entity._id,
+      if (deployDepsList === undefined) {
+        continue;
+      }
+    } else if (depsList.length > 0) {
+        for (const dep of depsList) {
+          const fullName = dep.substring(1, dep.indexOf(':'));
+          let depRepo = await j1Client.queryV1(
+            `FIND CodeRepo WITH full_name='${fullName}'`,
```
Comment (austin-rausch) [inline src/index.ts]: You should add some verification here to make sure that there is only one `CodeRepo` found.  E.G. add `LIMIT 1` to the query and also verify the length, if there are two I would think erroring is the right approach.

File: src/index.ts [hunk 13]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
+  console.log('');
+  const j1Client = await getClient(clientInput);
+  const repoMap = await getRepoIds(pathToRepos, clientInput);
+  console.log('');
 
-  for (const repoName of Array.from(repoMap.values())) {
+  for (const repoName of Array.from(repoMap.keys())) {
     console.log('Repo: ' + repoName);
 
     const mainRepo = await j1Client.queryV1(
```
Comment (austin-rausch) [inline src/index.ts:41]: You should add some verification here to make sure that there is only one CodeRepo found. E.G. add LIMIT 1 to the query and also verify the length, if there are two I would think erroring is the right approach.

File: src/index.ts [hunk 14]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
+            }
+          }
+        }
+      } else {
+        console.log('Repo has no dependencies of requested scopes');
+      }
+    if (deployDepsList != undefined) {
+      if (deployDepsList.length > 0) {
+        for (const dep of deployDepsList) {
+          let deployRepo = await j1Client.queryV1(
```
Comment (austin-rausch) [inline src/index.ts]: Same here as the above comments, work verifying only one result.

File: src/index.ts [hunk 15]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
           );
-          failure++;
-          if (missingDeps.indexOf(dep) === -1) {
-            missingDeps.push(dep + ' (' + repoName + ').');
+          if (deployRepo.length > 0) {
+            const relationshipKey = mainRepo[0].entity._key + '|depends_on|' + deployRepo[0].entity._key;
+            const relationship = await j1Client.createRelationship(
+              relationshipKey,
+              'bitbucket_repo_depends_on_bitbucketrepo',
+              'USES',
```
Comment (austin-rausch) [inline src/index.ts]: So we have `depends_on` listed in the type key but then have `USES` here for the relationship class, I don't think we have any two word classes.  Maybe `REQUIRES` or `NEEDS`? Probably worth getting Erkang's 2 cents on that.

File: src/index.ts [hunk 16]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
...
+            );
+            console.log('Successfully created relationship (' + repoName + ' USES ' + dep + ').');
+            success++;
+          } else {
+            console.log(
+              'Failed to create relationship with ' + dep + 
+              ' (was not found on the graph). Skipped.'
+            );
+            failure++;
+            if (missingDeps.indexOf(dep) === -1) {
```
Comment (austin-rausch) [inline src/index.ts]: You can use `.includes` here instead, it returns a boolean. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

File: src/index.ts [hunk 17]
```diff
@@ -9,80 +10,117 @@ require('dotenv').config();
 let success = 0, failure = 0;
 const missingDeps = [];
 let packageScope = [];
+let J1_ACCOUNT, J1_USERNAME, J1_PASSWORD;
 
 (async () => {
   while (true) {
-    var scope = readlineSync.question('Please input a package scope, i.e. @package (input QUIT when finished): ');
+    var scope = readlineSync.question('Input a package scope, i.e. @package (input QUIT when finished): ');
```
Comment (austin-rausch) [inline src/index.ts]: This should be `const` or `let`, we avoid `var` in all cases because of how it behaves poorly with some block scopes.

File: src/getdependencies.ts [hunk 18]
```diff
@@ -1,24 +1,44 @@
...
 let dependencies = [];
 
-export function getDependencies(repo: string, packageScope: string[]) {
+export function getDependencies(repo: string, packageScope: string[], repoMap: Map<string,string>) {
   try {
-    let path = '../repos/' + repo + '/package.json';
-
+    let path = '../' + repoMap.get(repo) + '/' + repo + '/package.json';
     dependencies = [];
     readJson = require(path).dependencies;
```
Comment (austin-rausch) [inline src/getdependencies.ts]: You should name this something better, E.G. `repoDependencies`

File: src/getdependencies.ts [hunk 19]
```diff
@@ -1,24 +1,44 @@
...
     for (const j in readJson) {
```
Comment (austin-rausch) [inline src/getdependencies.ts]: This should be named better, `i`, `j`, and the such are usually used for index arrays, E.G. `for (let i = 0; i < 10; i++) {`. Since what this is reading is the `package.json`'s dependencies you should be able to name this `dependencies`.

File: src/getdependencies.ts [hunk 20]
```diff
@@ -1,24 +1,44 @@
...
-      for (const scope of packageScope) {
-        if (j.includes(scope)) {
-          dependencies.push(j + ': ' + readJson[j]);
+      if (packageScope.length === 0) {
+        dependencies.push(j + ': ' + readJson[j]);
+      }
+      else {
+        for (const scope of packageScope) {
+          if (j.includes(scope)) {
```
Comment (austin-rausch) [inline src/getdependencies.ts]: `scope` is for things like `@lifeomic/` will match `@lifeomic/some-service`, right? If thats the case you should check for `.startsWith(scope)` instead, right now this would match with scope of `jupiter` would match `some-jupiter-moon`.

File: src/getdependencies.ts [hunk 21]
```diff
@@ -1,24 +1,44 @@
...
+            dependencies.push(j + ': ' + readJson[j]);
```
Comment (austin-rausch) [inline src/getdependencies.ts]: So this is setting a version after the `:` right? Usually this is denoted as `${packageName}@${versionSemVar}` E.G. `@lifeomic/attempt@^1.0.0`

**Review by austin-rausch — CHANGES_REQUESTED** [review]

File: src/getrepos.ts (hunk 11 above)
Comment (mnguo98) [reply src/getrepos.ts]: LMAO. I need to stop doing this since apparently I always forget to change it later.

File: src/getdependencies.ts (hunk 8 above)
Comment (mnguo98) [reply src/getdependencies.ts:32]: The console output will show the repo name followed by the operations.
```
Repo: graphql-tools-cli
*** Repo does not have a deploy directory with a dependencies.yaml ***

Successfully created relationship (graphql-tools-cli USES @lifeomic/alpha: ^0.14.0).
Failed to create relationship with @lifeomic/axios-fetch: ^1.4.0 (was not found on the graph). Skipped.
Successfully created relationship (graphql-tools-cli USES @lifeomic/platform-headers: ^1.3.2).
```

File: src/getdependencies.ts (hunk 21 above)
Comment (mnguo98) [reply src/getdependencies.ts]: I see. I was trying to keep it consistent with how the package.json shows it

File: src/getrepos.ts [hunk 22]
```diff
@@ -1,31 +1,36 @@
...
+  const repos = readdirSync(repoPath);
+  const j1Client = await getClient(clientInput);
 
   for (const repo of repos) {
     const repoID = await j1Client.queryV1(
       `Find CodeRepo with name='${repo}'`,
     );
     if (repoID.length === 0) {
-      console.log('Could not query Repo (' + repo + ').');
+      const newRepoPath = repoPath.substring(repoPath.length) === '/' ? repoPath + repo : repoPath + '/' + repo;
```
Comment (austin-rausch) [inline src/getrepos.ts]: `repoPath.substring(repoPath.length)` will always be an empty string.  If you're trying to get the last character you'll need to use `repoPath.length - 1` and you can also just index at the string like you would an array. E.G. `repoPath[repoPath.length - 1]`.

File: src/index.ts [hunk 23]
```diff
@@ -1,88 +1,106 @@
...
+  const password = question('Input password. If using a .env file, continue without input: ');
+  const access = question('Input access token. If using a .env file, continue without input: ');
+  const clientInput = {account: account, username: username, password: password, accessToken: access};
+
 
-  const repoMap = await getRepoIds();
+  console.log('');
+  const j1Client = await getClient(clientInput);
+  const repoMap = await getRepoIds(pathToRepos, clientInput);
+  console.log('');
```
Comment (austin-rausch) [inline src/index.ts:36]: Just making sure, are these console logs intentional to make output look nicer?

File: src/index.ts [hunk 24]
```diff
@@ -1,88 +1,106 @@
...
-          console.log(
-            'Failed to create relationship with ' + dep + 
-            ' (was not found on the graph). Skipped.'
-          );
-          failure++;
-          if (missingDeps.indexOf(dep) === -1) {
-            missingDeps.push(dep + ' (' + repoName + ').');
-          }
-        }
+    switch (switchName) {
```
Comment (austin-rausch) [inline src/index.ts]: I think you can simplify a lot of this, just have two if / else blocks.

```
if (depsList !== undefined) {
    const relationship = await createRepoRelationship(depsList, mainRepo, j1Client, repoName, missingDeps);
    missingDeps = relationship.missingDeps;
    success += relationship.success;
    failure += relationship.failure;
} else {
    console.log('No package.json, not creating repo dependency relationships')
}

if (deployDepsList !== undefined) {
    const relationship = await createDeployRelationship(deployDepsList, mainRepo, j1Client, repoName, missingDeps);
    missingDeps = relationship.missingDeps;
    success += relationship.success;
    failure += relationship.failure;
} else {
    console.log('No dependencies.yaml, not creating deploy relationships');
}
```

File: src/createRepoRelationship.ts [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    }
+  } else {
+    console.log('Repo has no dependencies of requested scopes');
+  }
+  successFail.missingDeps = missingDeps;
+  return successFail;
+}
+
+export async function createDeployRelationship(deployDepsList, mainRepo, j1Client, repoName, missingDeps) {
```
Comment (austin-rausch) [inline src/createRepoRelationship.ts]: Minor, but this function should probably be named plurally, E.G. `createDeployRelationships` since it creates multiple.

File: src/createRepoRelationship.ts [hunk 26]
```diff
@@ -0,0 +1,82 @@
+export async function createRepoRelationship(depsList, mainRepo, j1Client, repoName, missingDeps: string[]) {
```
Comment (austin-rausch) [inline src/createRepoRelationship.ts]: Minor, but this function should probably be named plurally, E.G. `createRepoRelationships` since it creates multiple.

**Review by austin-rausch — CHANGES_REQUESTED** [review]
Everything is looking good, only thing I'd like to see changed is the switch statement.

File: README.md [hunk 27]
```diff
@@ -1,11 +1,66 @@
-# jupiter-repo-dependencies
+# JupiterOne Repository Relationship Generator
 
-- Put the repos you want to run the script on in the `repos` folder.
-- Use `yarn deps` to run the script.
+A script that generates graphql relationships between a repository and its dependencies.
```
Comment (aiwilliams) [inline README.md]: I think you mean JupiterOne entity relationships.

File: README.md [hunk 28]
```diff
@@ -1,11 +1,66 @@
...
+This script will look for dependencies within a repository's `package.json` file as well as the `dependencies.yaml`file if the repository has a `deploy` directory.
 
-## Package Scope Input
+This script uses the JupiterOne graphql graph database.
```
Comment (aiwilliams) [inline README.md]: Strictly speaking, the JupiterOne graph database is distinct from the use of GraphQL in the API interfaces to the JupiterOne system. I think this sentence can be dropped.

File: README.md [hunk 29]
```diff
@@ -1,11 +1,66 @@
...
 
-- You will be asked to input as many package scopes as you like.
-- For example, if you have packages that follow this pattern: @package/*, then you would input @package.
+## Before Running
+- Create a directory containing the repositories that you would like the script to be run on anywhere you would like in the project (Note: You will need the path to this directory while running the script).
```
Comment (aiwilliams) [inline README.md]: I think "anywhere you would like in the project" can be dropped. It's a bit confusing because I thought it was suggesting I place that directory *within this project*.

File: README.md [hunk 30]
```diff
@@ -1,11 +1,66 @@
...
 
+- Ensure that the `package.json` for each repository is at the root of the repository.
+
+- JupiterOne Client credentials are required to run the necessary queries to create the relationships.
```
Comment (aiwilliams) [inline README.md]:
```suggestion
- JupiterOne API credentials are required to run the necessary queries to create the relationships.
```

File: README.md [hunk 31]
```diff
@@ -1,11 +1,66 @@
...
+  - There are 4 parameters that need to be passed in:
+    - Account
+    - Username
+    - Password
+    - AccessToken
```
Comment (aiwilliams) [inline README.md]: I've never looked closely, does the API require username/password AND an access token? Coming at this with a bit of inexperience, it seems surprising that I would have to give my username/password at all. Perhaps there could be a link to the JupiterOne API authentication documentation.

File: package.json [hunk 32]
```diff
@@ -4,22 +4,25 @@
   "version": "1.0.0",
   "license": "UNLICENSED",
   "scripts": {
-    "deps": "./tools/bin/run-create-relationships",
-    "deploy": "lifeomic-manual-deploy -p $npm_package_name -t lifeomic-test --action apply --skip-terraform"
+    "deps": "./tools/bin/run-create-relationships"
```
Comment (aiwilliams) [inline package.json]: `yarn deps` was surprising to me. I think `yarn create-relationships` is more obvious.

File: src/index.ts (hunk 23 above)
Comment (mnguo98) [reply src/index.ts:36]: Yeah they are

File: src/index.ts (hunk 24 above)
Comment (mnguo98) [reply src/index.ts]: Bro whatcha got against switch statements

File: src/getrepos.ts (hunk 22 above)
Comment (mnguo98) [reply src/getrepos.ts]: Oh yeah thanks

File: README.md (hunk 29 above)
Comment (mnguo98) [reply README.md]: That was my original intent but I guess it would work as long as the path was correct

Comment (aiwilliams) [reply README.md]: I don't like to put stuff in my working directory that I have to add to `.gitignore` 😁

**Review by austin-rausch — APPROVED** [review]
